# Import Libraries
#### Standard

In [81]:
import os
import numpy as np
import pandas as pd

#### Visualization

In [82]:
import matplotlib.pyplot as plt
import seaborn as sns

#### Visualization Setting

In [83]:
color_palette = 'Set3'
title_size = 20
sns.set_style("whitegrid")

# Import Data

In [84]:
root = os.path.join(os.getcwd(), 'DATA')

# 직원
jikwon_dir = os.path.join(root, 'jikwon.csv')
jikwon = pd.read_csv(jikwon_dir, encoding='cp949')

# 프로그램 변경 이력
change_dir_1 = os.path.join(root, 'program_change_1.xlsx')
change_dir_2 = os.path.join(root, 'program_change_2.xlsx')
program_change_1 = pd.read_excel(change_dir_1)
program_change_2 = pd.read_excel(change_dir_2)

# 프로젝트
project_dir = os.path.join(root, 'project.xlsx')
project = pd.read_excel(project_dir, sheet_name='프로젝트정보')
project_jikwon = pd.read_excel(project_dir, sheet_name='프로젝트인력')

# Data Overview
## Jikwon

In [85]:
print("Jikwon")
print(*jikwon.columns.values, sep=", ")

Jikwon
JIKWON_NO, NAME, JEOM_NO, JEOM_NAME, JIKGUN, JIKGEUB, JIKWHI, JIKWHI_NAME, JUMINNO, JIKWHI2, JUJKMU_C, JUJKMU_NM, JUJKMU_RATE, BUJKMU_C, BUJKMU_NM, BUJKMU_RATE, JUMJANG_G, HOBONG, TOIJIKGBN, BUIM_ILJA, LST_CHG_ILSI, SILNO_CID, CUSNO, SAME_MN_NO


In [86]:
jikwon.head()

JIKWON_NO               NAME  JEOM_NO                    JEOM_NAME  JIKGUN  \
0   20505190  최태호                      4    준법감시부                       30000   
1    6209769  김명희                   1107  영등포                           20004   
2   10601562  권수진                   7097    안동법원(출)                     40005   
3    8200858  신주아                   7051   강원도청                         20004   
4    8502374  김용섭                     19    사회공헌부                       40002   

   JIKGEUB  JIKWHI JIKWHI_NAME JUMINNO  JIKWHI2  ...  BUJKMU_NM BUJKMU_RATE  \
0    99999   20425      전문계약직원     NaN        9  ...        NaN         NaN   
1        3   31053   부지점장겸 WRM     NaN        2  ...        NaN         0.0   
2       22   31035          선임     NaN        9  ...        NaN         0.0   
3        4   20161          과장     NaN        2  ...    SOHO BM        20.0   
4        0   20456        사무직원     NaN        9  ...        NaN         0.0   

   JUMJANG_G  HOBONG TOIJIKGBN  BUIM_ILJA  LST_CHG_ILSI      SILNO_CID  \
0          0   99998         0   20200401           NaN  fLOjQav3YWgEC   
1          0   10975         0   20190128  2.019013e+13  yoV049wbj4xLo   
2          0   52235         0   20181210  2.019062e+13  j8QEZ81uQP0+I   
3          0   32055         0   20190710  2.021020e+13  1LxXLLTuM0YV2   
4          0   99998         0   20080707  6.172019e+12  qqu6RxZRXai1Z   

       CUSNO  SAME_MN_NO  
0  464464292   464464292  
1  478015769   478015769  
2  739479256   739479256  
3  626570106   626570106  
4  564647680   564647680  

[5 rows x 24 columns]

## Program

In [87]:
print("Program")
print(*program_change_1.columns.values, sep=", ")

Program
COL01, COL02, COL03, COL04, COL05, COL06, COL07, COL08, COL09, COL10, COL11, COL12, COL13, COL14, COL15, dw_lst_jukja_dt


Program 데이터의 경우 컬럼명이 실제 컬럼값을 대변하지 않는 것을 확인할 수 있는데, 컬럼값을 보면서 컬럼을 유추해야한다

In [88]:
program_change_1.head()

COL01             COL02             COL03        COL04  \
0        코어                수신             수신 공통        채널매핑룰   
1  프로프레임4.0         JPK-일본ROC             글로벌여신  프로프레임소스(.c)   
2        단위     단위(신)_BPR(AP)             BPR수신      ONL_OMM   
3      웹컨텐츠  문서중앙화시스템(Sdrive)  문서중앙화시스템(Sdrive)        Image   
4  프로프레임4.0      JBS-(신)일본ROC             글로벌공통  프로프레임소스(.c)   

                          COL05 COL06       COL07           COL08     COL09  \
0  A0_0003200002_SSDM0320A001_I   이창건  BK08101329  P2020-15765-01  20201228   
1                   sgle9999a.c   박창언  BK10101011  P2020-09188-02  20200818   
2         SBDP2305F702InDto.omm   김종배  BK15061008  P2020-08091-01  20200821   
3            btn_submit_del.gif   조아라  BK17092004  P2020-09726-01  20210401   
4    pfmDbiogif_trx_mas_s002.pc   이준현  BK18100783  P2020-05278-01  20200427   

  COL10           COL11                                       COL12  \
0    수정  2020-04-059662  /webdev/chl1/CHLDownload/trans/mappingrule   
1    신규  2020-04-042659                             /jpsdev/gln/src   
2    수정  2020-04-043394      /BprNbpBdp/src/bpr/nbp/bdp/service/dto   
3    신규  2021-04-011705    /skin/blue_en/images/content-page/button   
4    수정  2020-04-026378                            /jpsdev/dbio/src   

       COL13  COL14  COL15  dw_lst_jukja_dt  
0         코어      6      0         20210407  
1  DS_글로벌뱅킹3      1    695         20210407  
2     DS_BPR      3      1         20210407  
3   DS_경영정보1      1      0         20210407  
4     글로벌E뱅킹      2      0         20210407

- COL01: 시스템구분
- COL02: 
- COL03: 
- COL04: 프로그램 종류
- COL05: 프로그램명
- COL06: 변경자명
- COL07: 변경자행번
- COL08: 프로젝트번호
- COL09: 변경일자
- COL10: 변경구분
- COL11: 신청번호
- COL12: 경로
- COL13: 변경자부서
- COL14:
- COL15:
- COL16: DW적재일자

## Project

In [89]:
print("Project")
print(*project.columns.values, sep=", ")
print("\nProject Jikwon")
print(*project_jikwon.columns.values, sep=", ")

Project
프로젝트번호, 프로젝트명, 프로젝트 상태, PM행번, PM성명, PM부서, 업무명, 프로그램변경유무, 프로젝트계획시작일, 프로젝트계획종료일, 착수일, 종료일, 이행예정일, 이행일, 개발의뢰번호, 의뢰자행번, 의뢰자정보, 품질검증대상유무, 패스트체크인여부, 체크인가능유무, 인수테스트 완료여부, BA결재 여부, 인수테스트대상유무, 인수테스트 실시건수, 인수테스트 마지막 등록일, 통합테스트유무, Helpdesc공지대상유무, Helpdesc공지대상구분

Project Jikwon
프로젝트번호, 행번, 성명, 부서명, 팀명, 직위, 직급, 역할, 계획공수, 실제공수


In [90]:
project.head()

프로젝트번호                         프로젝트명 프로젝트 상태      PM행번 PM성명  \
0  P2018-1020-01              IBMS 지분투자 플랫폼 개선      종료  16031001  김현주   
1  P2018-6996-01    [캄보디아] CSS 연동을 통한 CD공동망 개발      종료   6127258  고규범   
2  P2018-6996-02  [신한캄보디아] CSS 연동을 통한 CD공동망 개발      종료   6181384  조원희   
3  P2018-6996-06     [크메르] CSS 연동을 통한 CD공동망 개발      종료  18101046  홍성근   
4  P2018-7039-01                新 신탁운용시스템 개발의뢰      종료   6150608  이재홍   

      PM부서           업무명 프로그램변경유무  프로젝트계획시작일  프로젝트계획종료일  ...  패스트체크인여부  \
0  DS_종금신탁  IB 프로젝트 기획관리        Y   20180206   20200930  ...         N   
1    글로벌수신        채널-Dex        Y   20180920   20200214  ...         Y   
2    글로벌수신        수신-Dex        Y   20181207   20200214  ...         Y   
3    글로벌수신        공통-Dex        Y   20190603   20200214  ...         N   
4       신탁        신탁유가증권        Y   20180921   20200720  ...         Y   

   체크인가능유무  인수테스트 완료여부  BA결재 여부 인수테스트대상유무  인수테스트 실시건수 인수테스트 마지막 등록일 통합테스트유무  \
0        Y           Y        N         Y           1    20190121.0       N   
1        Y           Y        N         Y           1    20200130.0       Y   
2        Y           Y        N         Y           1    20200115.0       Y   
3        Y           Y        N         Y           1    20200115.0       Y   
4        Y           Y        N         Y           1    20190628.0       Y   

  Helpdesc공지대상유무 Helpdesc공지대상구분  
0              N            NaN  
1              N            NaN  
2              N            NaN  
3              N            NaN  
4              N            NaN  

[5 rows x 28 columns]

In [108]:
project_jikwon.head()

프로젝트번호        행번   성명     부서명         팀명    직위  직급   역할  계획공수  \
0  P2018-1020-01  16031001  김현주  글로벌개발부  DS_글로벌뱅킹1  S부부장  기타   PM  1829   
1  P2018-1020-01  06172202  손현수   금융개발부         IB    수석  Ma  개발자     1   
2  P2018-1020-01  08101003  배준곤   DS개발팀     DS_BPR  S부부장  기타  개발자     0   
3  P2018-1020-01  18031010  박경규   DS개발팀     DS_수신3   S선임  기타  개발자   150   
4  P2018-6996-01  06127258  고규범  글로벌개발부     글로벌BRM    수석  4급   PM     0   

     실제공수  
0  2577.0  
1     1.0  
2    24.0  
3   197.5  
4     0.0


# Join Data
## Create Jikwon Index

In [91]:
jikwon.shape

(13973, 24)

- 24 Features, 13973 Entries

-> ICT 그룹 직원을 메인 인덱스로 사용하기 위해 점번호를 기준으로 필터링
- 61(ICT운영부)
- 62(정보개발부)
- 63(금융개발부)
- 64(글로벌개발부)
- 66(기관개발부)
- 69(디지털개발부)
- 507(ICT기획부)

In [92]:
main_data = jikwon[(jikwon['JEOM_NO'] == 61) | (jikwon['JEOM_NO'] == 62) | (jikwon['JEOM_NO'] ==  63) |
                   (jikwon['JEOM_NO'] == 64) | (jikwon['JEOM_NO'] == 66) | (jikwon['JEOM_NO'] == 69) | (jikwon['JEOM_NO'] ==  507)]

In [93]:
main_data.shape

(432, 24)

402명으로 직원 수를 줄였다

이를 기반으로 모든 데이터를 JOIN 할 것이다

## Merge Data
### Wrangling Program Data
데이터 조인에 앞서 프로그램 변경이력의 경우 2개의 dataframe으로 나뉘어져 있기 때문에 이를 합치도록 한다

In [94]:
program = pd.concat([program_change_1, program_change_2], axis=0)

In [95]:
print(program_change_1.shape)
print(program_change_2.shape)
print(program.shape)

(319998, 16)
(318125, 16)
(638123, 16)


또한, 직원 데이터의 직원번호는 **int형**인 반면, 프로그램 데이터의 직원번호는 'BK' + 행번 형태의 **object형**임을 이전에 확인했다

In [96]:
jikwon['JIKWON_NO']

0        20505190
1         6209769
2        10601562
3         8200858
4         8502374
           ...   
13968     8201943
13969    13600036
13970    19100840
13971    20100700
13972     6211895
Name: JIKWON_NO, Length: 13973, dtype: int64

In [97]:
program['COL07']

0         BK08101329
1         BK10101011
2         BK15061008
3         BK17092004
4         BK18100783
             ...    
318120    BK15112004
318121    BK20100167
318122    BK06226973
318123    BK20200821
318124    BK16111008
Name: COL07, Length: 638123, dtype: object

프로그램 데이터의 직원번호를 직원데이터와 형태가 일치하도록 수정한다

그 전에, 모든 직원번호가 **BK + 행번** 형태인지 확인하자

In [98]:
program[program['COL07'].str[:2] != 'BK'].COL07

619        SGO001348
716        SGO001338
968        SGO001338
1419      SGO1000511
3234       SGO001354
             ...    
316002     SGO001332
316142    SGO1000615
316540     SGO001347
317327     SGO001348
317900     SGO001354
Name: COL07, Length: 1107, dtype: object

이렇듯, BK로 시작하지 않는, 다시 말해 당행 직원이 아닌 데이터가 있다

해당 데이터는 현재 단계에서 필요하지 않기 때문에 제외하도록 한다

In [99]:
print(program.shape, end="->")
program = program[program['COL07'].str[:2] == 'BK']
print(program.shape)

(638123, 16)->(637016, 16)


이제 다시 직원데이터와 형태가 일치하도록 변경한다

In [100]:
program['COL07'] = program['COL07'].str[2:].astype(int)

In [101]:
program.head()

COL01             COL02             COL03        COL04  \
0        코어                수신             수신 공통        채널매핑룰   
1  프로프레임4.0         JPK-일본ROC             글로벌여신  프로프레임소스(.c)   
2        단위     단위(신)_BPR(AP)             BPR수신      ONL_OMM   
3      웹컨텐츠  문서중앙화시스템(Sdrive)  문서중앙화시스템(Sdrive)        Image   
4  프로프레임4.0      JBS-(신)일본ROC             글로벌공통  프로프레임소스(.c)   

                          COL05 COL06     COL07           COL08     COL09  \
0  A0_0003200002_SSDM0320A001_I   이창건   8101329  P2020-15765-01  20201228   
1                   sgle9999a.c   박창언  10101011  P2020-09188-02  20200818   
2         SBDP2305F702InDto.omm   김종배  15061008  P2020-08091-01  20200821   
3            btn_submit_del.gif   조아라  17092004  P2020-09726-01  20210401   
4    pfmDbiogif_trx_mas_s002.pc   이준현  18100783  P2020-05278-01  20200427   

  COL10           COL11                                       COL12  \
0    수정  2020-04-059662  /webdev/chl1/CHLDownload/trans/mappingrule   
1    신규  2020-04-042659                             /jpsdev/gln/src   
2    수정  2020-04-043394      /BprNbpBdp/src/bpr/nbp/bdp/service/dto   
3    신규  2021-04-011705    /skin/blue_en/images/content-page/button   
4    수정  2020-04-026378                            /jpsdev/dbio/src   

       COL13  COL14  COL15  dw_lst_jukja_dt  
0         코어      6      0         20210407  
1  DS_글로벌뱅킹3      1    695         20210407  
2     DS_BPR      3      1         20210407  
3   DS_경영정보1      1      0         20210407  
4     글로벌E뱅킹      2      0         20210407

## 직원 데이터 + 프로그램 변경 이력

In [103]:
main_data.shape

(432, 24)

In [104]:
main_data = pd.merge(left=main_data, right=program, how='left', left_on='JIKWON_NO', right_on='COL07')

In [107]:
main_data.shape

(370040, 40)

In [109]:
main_data.head()

JIKWON_NO               NAME  JEOM_NO                  JEOM_NAME  JIKGUN  \
0   20101491  신범철                     63  금융개발부                       20004   
1   20101491  신범철                     63  금융개발부                       20004   
2   20101491  신범철                     63  금융개발부                       20004   
3   20101491  신범철                     63  금융개발부                       20004   
4   20101491  신범철                     63  금융개발부                       20004   

   JIKGEUB  JIKWHI JIKWHI_NAME JUMINNO  JIKWHI2  ...       COL07  \
0        5   20164          대리     NaN        3  ...  20101491.0   
1        5   20164          대리     NaN        3  ...  20101491.0   
2        5   20164          대리     NaN        3  ...  20101491.0   
3        5   20164          대리     NaN        3  ...  20101491.0   
4        5   20164          대리     NaN        3  ...  20101491.0   

            COL08       COL09  COL10           COL11  \
0  P2021-00342-01  20210120.0     수정  2021-04-002421   
1  P2021-00342-01  20210120.0     수정  2021-04-002421   
2  P2021-00342-01  20210120.0     수정  2021-04-002421   
3  P2021-01486-01  20210209.0     수정  2021-04-005149   
4  P2021-01319-02  20210317.0     수정  2021-04-008846   

                                               COL12  COL13  COL14  COL15  \
0                                /nbsdev/fib/fis/src   지불결제    6.0    4.0   
1                                /nbsdev/fib/fis/src   지불결제   10.0   15.0   
2  /webdev/chl1/CHLDownload/iWorks/DOWNLOAD/res/b...   지불결제   39.0    2.0   
3                                /nbsdev/via/vas/src   지불결제    3.0  105.0   
4                                /nbsdev/uif/elb/src   지불결제   97.0   53.0   

   dw_lst_jukja_dt  
0       20210407.0  
1       20210407.0  
2       20210407.0  
3       20210407.0  
4       20210407.0  

[5 rows x 40 columns]

## 직원데이터 + 프로젝트 정보
프로그램의 프로젝트번호를 기반으로 조인한다

In [110]:
main_data = pd.merge(left=main_data, right=project, how='left', left_on='COL08', right_on='프로젝트번호')

In [111]:
main_data = pd.merge(left=main_data, right=project_jikwon, how='left', left_on='COL08', right_on='프로젝트번호')

In [124]:
main_data.describe(include=['O'])

NAME                 JEOM_NAME JIKWHI_NAME JUMINNO  \
count             2705545                   2705545     2705545       0   
unique                425                         7          11       0   
top     이현구                디지털개발부                           부부장     NaN   
freq              1305282                   1546468     1398624     NaN   

       JUJKMU_NM BUJKMU_NM      SILNO_CID    COL01          COL02    COL03  \
count    2702668     15767        2705545  2705424        2705424  2705424   
unique        65         7            432        4            283      402   
top         여신개발     모바일개발  VcFt8DqiqAIY0       코어  (U2L) 단위-CCRM     신용평가   
freq     1433049     12830        1305282  1530477        1335333   430234   

        ... Helpdesc공지대상유무 Helpdesc공지대상구분        프로젝트번호_y        행번       성명  \
count   ...        2705424         111842         2705424   2705424  2705424   
unique  ...              2              9            6691       541      531   
top     ...              N             여신  P2020-03928-01  10100806      박준원   
freq    ...        2624619          32770         1282060    111415   111415   

            부서명       팀명       직위       직급       역할  
count   2705424  2705424  2705424  2704163  2705424  
unique       58      122       36        6        2  
top      디지털개발부     기업여신       선임       5급      개발자  
freq    1417898   897574  1485700  1348411  2335480  

[4 rows x 44 columns]

In [131]:
print(*main_data.columns.values, sep="\n")

JIKWON_NO
NAME
JEOM_NO
JEOM_NAME
JIKGUN
JIKGEUB
JIKWHI
JIKWHI_NAME
JUMINNO
JIKWHI2
JUJKMU_C
JUJKMU_NM
JUJKMU_RATE
BUJKMU_C
BUJKMU_NM
BUJKMU_RATE
JUMJANG_G
HOBONG
TOIJIKGBN
BUIM_ILJA
LST_CHG_ILSI
SILNO_CID
CUSNO
SAME_MN_NO
COL01
COL02
COL03
COL04
COL05
COL06
COL07
COL08
COL09
COL10
COL11
COL12
COL13
COL14
COL15
dw_lst_jukja_dt
프로젝트번호_x
프로젝트명
프로젝트 상태
PM행번
PM성명
PM부서
업무명
프로그램변경유무
프로젝트계획시작일
프로젝트계획종료일
착수일
종료일
이행예정일
이행일
개발의뢰번호
의뢰자행번
의뢰자정보
품질검증대상유무
패스트체크인여부
체크인가능유무
인수테스트 완료여부
BA결재 여부
인수테스트대상유무
인수테스트 실시건수
인수테스트 마지막 등록일
통합테스트유무
Helpdesc공지대상유무
Helpdesc공지대상구분
프로젝트번호_y
행번
성명
부서명
팀명
직위
직급
역할
계획공수
실제공수
